presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergMetadata.java

```diff
@@ -521,4 +529,29 @@ public void rollback()
     {
         // TODO: cleanup open transaction
     }
+
+    @Override
+    public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle handle, Constraint constraint)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) handle;
+        TupleDomain<HiveColumnHandle> newDomain = convertTupleDomainTypes(constraint.getSummary().transform(HiveColumnHandle.class::cast));
+
+        if (newDomain.equals(table.getEffectivePredicate())) {
+            return Optional.empty();
+        }
+
+        List<HiveColumnHandle> partitionColumns = fetchParititonColumns(getIcebergTable(metastore, hdfsEnvironment, session, table.getSchemaTableName()), getColumnHandles(session, table));
+
+        TupleDomain<HiveColumnHandle> unenforcedTupleDomain = TupleDomain.withColumnDomains(Maps.filterKeys(newDomain.getDomains().get(), not(Predicates.in(partitionColumns))));
+        TupleDomain<HiveColumnHandle> enforcedTupleDomain = TupleDomain.withColumnDomains(Maps.filterKeys(newDomain.getDomains().get(), Predicates.in(partitionColumns)));
```

Comment:
In case of partition spec evolution, it is possible that `enforcedTupleDomain` is not enforced on files with old partition specs. ([DataTableScan::planFiles()](https://github.com/apache/incubator-iceberg/blob/master/core/src/main/java/org/apache/iceberg/DataTableScan.java#L71-L104))

As a result, it seems inaccurate to
1. pass `unenforcedTupleDomain` as `remainingFilter` to `ConstraintApplicationResult`
2. pass `table.getEffectivePredicate()` to `createParquetPageSource`

A high-level suggestion would be to consider using residual expression returned by `FileScanTask::residual()` in Iceberg. But I'm not sure if it is straightforward to convert `Expression` in Iceberg to `TupleDomain` in Presto.

Reply:
But doesn't Iceberg ensure strict filtering is applied for partition columns ?

Reply:
Maybe we can write a `ExpressionTranslator` which would convert Iceberg expression to `TupleDomain`. I'll update along with it

Reply:
Yes. Thanks for investigating on this.

So the idea was: suppose we have a table whose partition column evolving from `col1` to `col2`. As a result, some old files are partitioned on `col1` and newer files on `col2`. When we do a SELECT filtered on `col2`, old files can't be filtered in or out.

Reply:
Even if we wrote an expressionTranslator the issue is  in case of iceberg the residual is at file level and not at table level so what we really need is a way to represent the push down at split level and not at table level, unless we take up planning at this stage which is not a good idea. The table level push down that presto allows us to provide right now can only work if we can either verify the partition spec is not evolved or it has only evolved in a more inclusive manner, i.e from a partition by (date,hour) -> partition(date).

Reply:
Thanks for your insight. Then I guess we implement a simple filter push down with no feedback from the Connector on which expressions are enforced by the connector. 

Is there is any way to determine this condition

> partition spec is not evolved or it has only evolved in a more inclusive manner, i.e from a partition by (date,hour) -> partition(date).

If we didn't remove the partition filter then deleting a partition in Iceberg from Presto might not be supported as `ConnectorMetadata#applyDelete` has a pattern of `DeleteNode` -> `TableScanNode` which might happen if we remove the filter that will be applied on the partition columns.